[Keyboard] Nifty Numpad

Acliad opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->


<!--- This template is entirely optional and can be removed, but is here to help both you and us. -->
<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description
Add support for [Nifty Numpad](https://gitlab.com/Acliad/nifty-numpad) macropad

<!--- Describe your changes in detail here. -->

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] ~My change requires a change to the documentation.~
- [ ] ~I have updated the documentation accordingly.~
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] ~I have added tests to cover my changes.~
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

Acliad commented on keyboards/nifty_numpad/nifty_numpad.h:
```diff
@@ -0,0 +1,66 @@
+#ifdef RGB_IDLE_ENABLE
+#define _LAYER_RGB 15
+// Custom keycode declarations
+enum custom_keycodes {
```
Got it, update pushed.

Acliad commented on keyboards/nifty_numpad/nifty_numpad.h:
```diff
@@ -0,0 +1,66 @@
+
+#include "quantum.h"
+
+// min/max macros
```
Ope, I don't know why I didn't even think to check. I've updated my instances to use QMK's.

drashna commented on keyboards/nifty_numpad/keymaps/cadence/rules.mk:
```diff
@@ -0,0 +1,6 @@
+TAP_DANCE_ENABLE=yes
+RGB_IDLE_ENABLE=yes
+
+ifeq ($(strip $(RGB_IDLE_ENABLE)), yes)
+    OPT_DEFS += -DRGB_IDLE_ENABLE
+endif
```
Might make more sense to have this as a config.h define. Otherwise, this block should be moved to the keyboard's post_rules.mk
```suggestion
```

Acliad replied:
I generally like the idea of controlling it as a config, but I think I prefer to keep it enabled/disabled at the keymap level. I've moved the lines to post_rules.mk and deleted them from rules.mk

drashna commented on keyboards/nifty_numpad/keymaps/idle_rgb_example/rules.mk:
```diff
@@ -0,0 +1,6 @@
+TAP_DANCE_ENABLE=yes
+RGB_IDLE_ENABLE=yes
+
+ifeq ($(strip $(RGB_IDLE_ENABLE)), yes)
+    OPT_DEFS += -DRGB_IDLE_ENABLE
+endif
```
same
```suggestion
```

drashna commented on keyboards/nifty_numpad/nifty_numpad.c:
```diff
@@ -0,0 +1,247 @@
+}
+#endif
+
+void matrix_scan_kb(void) {
```
```suggestion
void housekeeping_task_kb(void) {
```

Acliad replied:
Oh, that makes sense. Done.

drashna commented on keyboards/nifty_numpad/nifty_numpad.c:
```diff
@@ -0,0 +1,247 @@
+    }
+    #endif
+
+    matrix_scan_user();
```
```suggestion
```

Acliad replied:
Done as part of change to housekeeping_task_kb()

drashna commented on keyboards/nifty_numpad/nifty_numpad.c:
```diff
@@ -0,0 +1,247 @@
+    rgb_matrix_state_t idle_rgb_matrix;
+} kb_config_t;
+
+static uint32_t idle_timer = 0; 
```
instead of using your own timer, you could use `last_input_activity_elapsed()` as this is used/updated by the core based on any activity (matrix scan, encoders, pointing devices, etc)

This would remove some of the necessity for the custom code you have.

Acliad replied:
That's much cleaner, thanks for the heads up! Implemented.

drashna commented on keyboards/nifty_numpad/nifty_numpad.c:
Could you run a format pass on this file (`qmk format-c`)

Acliad replied:
Done.

drashna commented on keyboards/nifty_numpad/nifty_numpad.h:
```diff
@@ -0,0 +1,52 @@
+};
+
+#ifdef RGB_IDLE_ENABLE
+#define _LAYER_RGB 15
```
You could probably default to (MAX_LAYER-1), as that would be ... accurate based on the layer state size. (for instance, via only supports 4 layers, by default and sets the max layer to whatever the layer count is set to)
https://github.com/qmk/qmk_firmware/blob/master/quantum/action_layer.h#L49-L70

Acliad replied:
Once again much cleaner, tested and works great. Thank you!

drashna commented on keyboards/nifty_numpad/readme.md:
```diff
@@ -0,0 +1,35 @@
+# Nifty Numpad
+
+![Nifty Numpad](https://gitlab.com/Acliad/nifty-numpad/-/raw/main/Docs/Render%20(Blue,%20SS%20screws).png)
```
Image not found? 

Also, please rehost on Imgur or similar hosting site.

Acliad replied:
That's odd, I tested it and that repo isn't private. Regardless, I've rehosted it on Imgur and updated the readme.

Acliad commented:
Addressed comments up to now. Note that I snuck in a hotfix (I hope that's okay). Commit 1a05228 has details, but I had inconsistent definitions for RGB_LAYER between nifty_numpad.c and the keymaps. Thanks a bunch for your time and suggestions @drashna.

drashna commented on keyboards/nifty_numpad/nifty_numpad.h:
```diff
@@ -0,0 +1,52 @@
+*/
+
+enum custom_keycodes {
+    RGB_DEF = SAFE_RANGE,
```
```suggestion
    RGB_DEF = QK_KB,
```

`SAFE_RANGE` is mapped to `QK_USER`, and both the kb and user keycodes have their own block.

Acliad replied:
Did not know that, thanks! Updated

drashna commented on keyboards/nifty_numpad/keymaps/cadence/keymap.c:
```diff
@@ -0,0 +1,270 @@
+
+// Custom keycode declarations
+enum cadence_keycodes {
+    EDIT_GEN = RGB_IEF + 1,
```
```suggestion
    EDIT_GEN = QK_USER,
```

Acliad replied:
Ditto to above.

drashna commented on keyboards/nifty_numpad/keymaps/cadence/keymap.c:
This looks to be a user keymap? 

> qmk/qmk_firmware is no longer accepting PRs for keymaps other than for manufacturer-supported keymaps. User keymap workflow has been documented [here](https://docs.qmk.fm/#/newbs) for several years. This change is to progressively reduce the maintenance burden on the project, and to allow us to focus on the core features of QMK.

Acliad replied:
I'm not sure if it counts as one–it at least wasn't intended to be. I intended to leave it as an example of how I've mapped the numpad for use in an ECAD program (Cadence Allegro, in this case). It's not tied back to a `/users/<name>`. Technically I'm the manufacturer and support the keymap. However, I can see how it would be a user keymap in the sense that it's essentially my personal map.

I do want to keep the repo as clean as possible. I didn't update it on this commit just to make sure I'm not missing something. If you still think it should be removed from this PR as a user keymap, I'm happy do that update. Thanks!

drashna replied:
It might help to clarify things if you add a readme to this folder, then.

Acliad replied:
It is done :)

drashna commented on keyboards/nifty_numpad/keymaps/default/rules.mk:
This file can be removed

Acliad replied:
Removed

drashna commented on keyboards/nifty_numpad/keymaps/idle_rgb_example/keymap.c:
```diff
@@ -0,0 +1,179 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include QMK_KEYBOARD_H
+#include "nifty_numpad.h"
+#include "quantum.h"
```
Shouldn't be needed.  `QMK_KEYBOARD_H` should resolve to `nifty_numpad.h` for this keyboard, and that file should have quantum.h included.
```suggestion
```

Acliad replied:
Updated

drashna commented on keyboards/nifty_numpad/readme.md:
```diff
@@ -0,0 +1,35 @@
+# Nifty Numpad
+
+![Nifty Numpad](https://i.imgur.com/jfB8tfK.png)
```
```suggestion
![Nifty Numpad](https://i.imgur.com/jfB8tfKh.png)
```

Acliad replied:
Added

Acliad commented:
Requested updates up to now have been addressed. Thanks again @drashna!

Acliad commented:
Hey, just checking if this can get reviewed. I know this is a passion project so I'm not trying to be pushy, just waiting on this merge to release my product :)

Thanks!

drashna commented on keyboards/nifty_numpad/config.h:
```diff
@@ -0,0 +1,18 @@
+
+#define EECONFIG_KB_DATA_SIZE 11
+
+// RGB Matrix Parameters
+#define RGB_MATRIX_DEFAULT_MODE RGB_MATRIX_SOLID_COLOR // Sets the default mode, if none has been set
+#define RGB_MATRIX_DEFAULT_SPD 100 // Sets the default animation speed, if none has been set
+#define RGB_MATRIX_MAXIMUM_BRIGHTNESS 255
+
+#define RGB_MATRIX_DEFAULT_HUE 127 // Sets the default hue value, if none has been set
+#define RGB_MATRIX_DEFAULT_SAT 255 // Sets the default saturation value, if none has been set
+#define RGB_MATRIX_DEFAULT_VAL RGB_MATRIX_MAXIMUM_BRIGHTNESS // Sets the default brightness value, if none has been set
```
These can all be migrated to info.json.  `qmk migrate` or `qmk info -kb (name) -f json` can assist with that.

drashna commented on keyboards/nifty_numpad/nifty_numpad.c:
```diff
@@ -0,0 +1,225 @@
+    uint8_t mode;
+    uint8_t speed;
+    HSV     hsv;
+} rgb_matrix_state_t;
```
should add some asserts to ensure the sizes are consistent/don't change. 

```suggestion
} rgb_matrix_state_t;

_Static_assert(sizeof(rgb_matrix_state_t) == 5, "Invalid size for rgb_matrix_state_t");

```

drashna commented on keyboards/nifty_numpad/nifty_numpad.c:
```diff
@@ -0,0 +1,225 @@
+typedef struct {
+    rgb_matrix_state_t active_rgb_matrix;
+    rgb_matrix_state_t idle_rgb_matrix;
+} kb_config_t;
```
```suggestion
} kb_config_t;

_Static_assert(sizeof(kb_config_t) == 10, "Invalid size for kb_config_t");

```

Acliad replied:
Assert added to both structs

Acliad commented:
Latest change requests addressed. Thanks again!

Acliad commented:
Is there anything else needed from me for this merge? 

Thanks!
Isaac

fauxpark commented on keyboards/nifty_numpad/readme.md:
````diff
@@ -0,0 +1,35 @@
+Enter the bootloader in 2 ways:
+
+* **Bootmagic reset**: Hold down the key at (0,0) in the matrix (top left) and plug in the keyboard
+* **Physical reset button**: Press the ```RST``` button while holding the ```BOOT_SEL``` button on the back of the PCB
````
```suggestion
* **Physical reset button**: Press the `RST` button while holding the `BOOT_SEL` button on the back of the PCB
```

Acliad replied:
Thanks @fauxpark. I've made the readme.md updates suggested. I also removed the Cadence Keymap. The others I'd like to release as part of the product.

fauxpark reviewed (changes requested):
User keymaps are no longer accepted into the repo. Please remove all of the custom code.

zvecr commented on keyboards/nifty_numpad/info.json:
```diff
@@ -0,0 +1,153 @@
+    "processor": "RP2040",
+    "rgb_matrix": {
+        "driver": "ws2812",
+        "max_brightness": 255,
+        "default_val": 10,
```
```suggestion
```

zvecr commented on keyboards/nifty_numpad/info.json:
```diff
@@ -0,0 +1,153 @@
+            "rainbow_moving_chevron": true,
+            "rainbow_pinwheels": true,
+            "raindrops": true,
+            "solid_color": true,
```
```suggestion
```

Acliad replied:
@zvecr Both comments addressed and pushed. Thanks!

Acliad commented:
I noticed the previous build and lint failed on keyboards that aren't mine and I didn't change files in. I've merged all the latest changes and recommitted. I'm not sure why that would have happened though. My keyboard passed the local lint and build tests.

fauxpark commented:
Seems like you've pulled in a bunch of other commits somehow, possibly through rebasing. I'm going to close this and request you resubmit it with only your commits. Also, you do not need to make any merge commits; this is only necessary if there are conflicts vs. the base branch. Since you're only adding a keyboard that should not be an issue.
